### DISCUSSION TITLE
Keyboard Shortcuts Suggestions

### DISCUSSION THREAD

**Discussion by DarkArc** [standardnotes/forum#2617] [Suggestions]
## A Reasonable Base

One of the best things I've seen in the realm of keyboard shortcuts is KDE's Global Shortcuts window:
![image](https://user-images.githubusercontent.com/778012/79394461-893a6d00-7f45-11ea-8b22-cb7267798285.png)

I could see a similar configuration pop over for keyboard shortcuts in standard notes. If you can imagine, various things that need keyboard shortcuts could programmatically register with standard notes, and then be displayed on the left (where application launchers, and other shortcuts are displayed). 

The sections could instead be "Global Shortcuts" (i.e. bindings to control the base Standard Notes, and global extensions like folders, theming extensions, etc) and "Note Bound Shortcuts" (i.e. bindings to control editors, and utilities related to editors). Alternatively, the sections could be broken down by component type ("Actions", "Core", "Editors", etc).

Realistically, it would also be nice to set a shortcut to some kind of "Default binding" for an "action type" (not "Action" in the Standard Notes sense)... 
![image](https://user-images.githubusercontent.com/778012/79395031-c94e1f80-7f46-11ea-8590-3ea1da60b6dd.png)

If you could image, during an editors registration of its keyboard shortcuts, it could provide a dummy value that means, "whatever the default bold shortcut is" for its own bold shortcut. This could manifest in the UI as each section (e.g. "Note Bound Shortcuts") have an entry that's just "Defaults". 

## Simplicity Through API

I understand, and respect there's a desire to keep things simplistic. I think this is a reasonable "base" for a keyboard shortcut system. However, I could see arguments to the contrary. With respect to that, I could see it being reasonable to implement this as a sort of UI independent protocol. 

At an API level something like the following could be provided:
```
let registrationObject = shortcutManager.registerKeybinding(
  /*name or thing registering the shortcut*/ "my-namespace:my-component", 
  /*shortcut name*/"editor.enbolden", 
  /*fallback if not provided*/"C-b",
  /*action to call*/() => { }
);

registrationObject.something(); // whatever ends up being need for extension developers if anything
```

The call to `registerKeybinding` could internally inside of `shortcutManager` create a structure:
```
{
  "my-namespace:my-component": [
    {
      "node": editor.enbolden",
      "fallback": "C-b",
      "action": () => { } 
    },
    /* ... some other bindings ... */
  ],
  /* ... some other components ... */
}
```

This structure managed by `shortcutManager` could then be read by an extension which provides and manages shortcuts. In effect, the core shortcut system merely keeps track of the requested keybindings, and provides API to implement keyboard management (e.g. something to notify the keybinding management extension that "C-b" was pressed).

It would then be up to the extension to determine what shortcuts to apply where (most likely via asking question like, what editor is focused?, what actions are relevant?, etc.), and how to determine and apply the user's requested keybinds.

This could provide the flexibility to do something like I've suggested above, and the flexibility to go further, and do potentially crazy things (if a user so desired), like having "C-b" switch to the note to the markdown pro editor, and create bold text. The sky is the limit, but the sky is the limit in a very contained way.

## The Need For "Monolithic" Keybinding Management

I think the keybinding manger would have to be monolithic and inaccessible by other extensions for practical purposes. While it might be cool for someone to have several keybinding systems all in play at the same time, this is likely to quickly lead to serious incompatibilities and headaches, not to mention, if an theme extension were to say it's a keybinding service, it could create cool effects as you type, but it could also log your keys, and I just don't think that's cool. In effect, this means keybinding extensions behave like editors, only one can really be active at any given time, and I think that's fine. 

The fancy behaviors you could get from a more robust monolithic builtin keybinding manager that registers multiple keybinding extensions is likely beyond overkill, and sources of serious headache and maintainability issues. (If really desired, maybe someone could build a keybinding extension off of this core system that did allow its own version of extensions to customize it? Thus, keeping standard notes itself simple)

## Other Considerations

- Touch Screen Gesture Management
- Mouse Gesture Management

I think reasonably these would be handled similarly, but independently. (e.g. I can have 1 touch screen gesture management extension, AND 1 keyboard shortcut management extension, AND 1 mouse gesture management extension)

- Platform Specific Keybindings (Moving from Mac OS -> Linux -> Windows)

This one would be tricky. It's potentially resolvable as emacs resolves it (i.e. `C` can mean different things on different platforms).

----

Very much curious as to the Standard Notes Team and community's thoughts on these issues. I hope I've brought something new and thoughtful to the discussion (existing discussion seemed to be "I want this", so I wanted to be a bit more specific of how I really want something like this to work/how I could see it working). 

I haven't looked much at how extensions really work inside of standard notes, so I apologize if there is some architectural reason that some of this may not be desirable, or something about this model doesn't fit. To be clear, this is intended as a rough idea for purposes of discussion, and not a request for a concrete design.

**Comment by fineseasalt** [comment]
Thank you for your feedback @DarkArc! I really appreciate you going into detail on this one.

We've been looking into shortcuts in the web/desktop apps lately and we definitely see them as a feature that could provide great value to some of our users. I can confirm that we do have plans to bring them in Standard Notes sooner or later (hopefully sooner though, around v4.0)

We're considering having a simple 'shortcuts settings' space built into Standard Notes that would probably come with certain limitations. I can imagine, that in the first version, for example, one could view but might not be able to change the key combinations. And that's exactly why I like your idea of opening our options up through an API. Will definitely pass your ideas along to the engineers as well.

Could you please tell me the top 3 specific shortcuts use-cases/actions that you, personally, could benefit from? Would be great if you could also provide me with some context :)

Thanks again!

**Comment by DarkArc** [comment]
> Could you please tell me the top 3 specific shortcuts use-cases/actions that you, personally, could benefit from? Would be great if you could also provide me with some context :)

Supporting this workflow would be great:

Jump to search/search for note -> select note -> edit note

This could work as `esc` takes you to the search, you type your text, and then use the arrow keys to navigate the results. This could also work as `esc` then use arrow keys to navigate the notes, maybe double `esc` clears the current search is there is one.

As an example of something I think does this well, download Telegram Desktop. It's basically the same thing but with messaging instead of notes.

I could also see a way to select tags using only the keyboard being helpful though I don't have an idea there for what shortcuts to use.

The second would have to be "create a new note". Just a `ctrl + n` would probably be intuitive enough there, or a `ctrl + t` (though if you ever add tabs, that would be confusing). 

The third one I would have to be, easy search within a note. Some kind of `ctrl + f` equivalent at least for most editors.

**Comment by AlJohri** [comment]
I wanted to throw in that the shortcut that I am missing most is to show and hide the sidebar (i.e. distraction free mode). In Visual Studio Code the shortcut is `⌘B` while in Sublime Text `⌘KB` to hide the sidebar. Also Notion uses `⌘\` to toggle the sidebar.

**Comment by myreli** [comment]
Relates to https://github.com/standardnotes/forum/discussions/2567
